oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3RequestDecorator.java

```diff
@@ -20,34 +20,64 @@
 import java.util.Properties;
 
 import com.amazonaws.services.s3.model.CopyObjectRequest;
+import com.amazonaws.services.s3.model.GeneratePresignedUrlRequest;
+import com.amazonaws.services.s3.model.InitiateMultipartUploadRequest;
 import com.amazonaws.services.s3.model.ObjectMetadata;
 import com.amazonaws.services.s3.model.PutObjectRequest;
+import com.amazonaws.services.s3.model.SSEAlgorithm;
+import com.amazonaws.services.s3.model.SSEAwsKeyManagementParams;
+import com.amazonaws.util.StringUtils;
 
 /**
  * This class to sets encrption mode in S3 request.
  *
  */
 public class S3RequestDecorator {
     DataEncryption dataEncryption = DataEncryption.NONE;
+    Properties props;
+    SSEAwsKeyManagementParams sseParams;
 
     public S3RequestDecorator(Properties props) {
         if (props.getProperty(S3Constants.S3_ENCRYPTION) != null) {
             this.dataEncryption = dataEncryption.valueOf(props.getProperty(S3Constants.S3_ENCRYPTION));
+
+            if (props.getProperty(S3Constants.S3_ENCRYPTION).equals(S3Constants.S3_ENCRYPTION_SSE_KMS)) {
+                String keyId = props.getProperty(S3Constants.S3_SSE_KMS_KEYID);
+                if(!StringUtils.isNullOrEmpty(keyId)){
+                    sseParams = new SSEAwsKeyManagementParams(keyId);
+                    sseParams.withAwsKmsKeyId(keyId);
+                } else {
+                    sseParams = new SSEAwsKeyManagementParams();
+                }
+            }
         }
     }
 
     /**
      * Set encryption in {@link PutObjectRequest}
      */
     public PutObjectRequest decorate(PutObjectRequest request) {
+        ObjectMetadata metadata;
         switch (getDataEncryption()) {
             case SSE_S3:
-                ObjectMetadata metadata = request.getMetadata() == null
+                metadata = request.getMetadata() == null
                                 ? new ObjectMetadata()
                                 : request.getMetadata();
                 metadata.setSSEAlgorithm(ObjectMetadata.AES_256_SERVER_SIDE_ENCRYPTION);
                 request.setMetadata(metadata);
                 break;
+            case SSE_KMS:
+                metadata = request.getMetadata() == null
+                                ? new ObjectMetadata()
+                                : request.getMetadata();
```

Comment:
This code duplicates the same code in the case statement above (for SSE_S3).  It seems like this could be applied to line 60 above, where ObjectMetadata is first declared, since it applies to all cases.  Then just set the algorithm as appropriate in the case statement.